ccl/changefeedccl: TestCatchupScanOrdering failed under stress

SHA: https://github.com/cockroachdb/cockroach/commits/205330426e50b966ede8a4644d77570d8265fe66

Parameters:
```
TAGS=
GOFLAGS=-race -parallel=2
```

To repro, try:

```
# Don't forget to check out a clean suitable branch and experiment with the
# stress invocation until the desired results present themselves. For example,
# using stress instead of stressrace and passing the '-p' stressflag which
# controls concurrency.
./scripts/gceworker.sh start && ./scripts/gceworker.sh mosh
cd ~/go/src/github.com/cockroachdb/cockroach && \
stdbuf -oL -eL \
make stressrace TESTS=TestCatchupScanOrdering PKG=github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl TESTTIMEOUT=5m STRESSFLAGS='-maxtime 20m -timeout 10m' 2>&1 | tee /tmp/stress.log
```

Failed test: https://teamcity.cockroachdb.com/viewLog.html?buildId=1602175&tab=buildLog

```

I191121 06:47:11.419197 129677 sql/event_log.go:132  [n1,client=127.0.0.1:45972,user=root] Event: "create_table", target: 53, info: {TableName:d.public.bank Statement:CREATE TABLE IF NOT EXISTS bank (id INT8 PRIMARY KEY, balance INT8, payload STRING, FAMILY (id, balance, payload)) User:root}
I191121 06:47:11.432933 129954 storage/replica_command.go:413  [n1,split,s1,r24/1:/{Table/28-Max}] initiating a split of this range at key /Table/53 [r25] (zone config)
I191121 06:47:11.459476 129759 workload/workloadsql/dataload.go:135  imported bank (0s, 10 rows)
I191121 06:47:11.491616 129677 sql/event_log.go:132  [n1,client=127.0.0.1:45972,user=root] Event: "set_cluster_setting", target: 0, info: {SettingName:kv.range_merge.queue_enabled Value:false User:root}
I191121 06:47:11.498003 129759 workload/workloadsql/workloadsql.go:97  starting 9 splits
I191121 06:47:11.503944 129677 storage/replica_command.go:413  [n1,s1,r25/1:/{Table/53-Max}] initiating a split of this range at key /Table/53/1/5 [r26] (manual)
I191121 06:47:11.577576 129677 storage/replica_command.go:413  [n1,s1,r26/1:/{Table/53/1/5-Max}] initiating a split of this range at key /Table/53/1/8 [r27] (manual)
I191121 06:47:11.616872 130499 storage/replica_command.go:413  [n1,s1,r25/1:/Table/53{-/1/5}] initiating a split of this range at key /Table/53/1/3 [r28] (manual)
I191121 06:47:11.617273 129677 storage/replica_command.go:413  [n1,s1,r27/1:/{Table/53/1/8-Max}] initiating a split of this range at key /Table/53/1/9 [r29] (manual)
I191121 06:47:11.655785 130072 storage/replica_command.go:413  [n1,s1,r26/1:/Table/53/1/{5-8}] initiating a split of this range at key /Table/53/1/7 [r30] (manual)
I191121 06:47:11.663228 130499 storage/replica_command.go:413  [n1,s1,r28/1:/Table/53/1/{3-5}] initiating a split of this range at key /Table/53/1/4 [r31] (manual)
I191121 06:47:11.672384 129677 storage/replica_command.go:413  [n1,s1,r25/1:/Table/53{-/1/3}] initiating a split of this range at key /Table/53/1/2 [r32] (manual)
I191121 06:47:11.699201 130072 storage/replica_command.go:413  [n1,s1,r26/1:/Table/53/1/{5-7}] initiating a split of this range at key /Table/53/1/6 [r33] (manual)
I191121 06:47:11.717271 129677 storage/replica_command.go:413  [n1,s1,r25/1:/Table/53{-/1/2}] initiating a split of this range at key /Table/53/1/1 [r34] (manual)
I191121 06:47:12.585193 130922 ccl/changefeedccl/changefeed_processors.go:659  [n1,client=127.0.0.1:45994,user=root] sinkless feed span /Table/53/1{-/1} is behind by 437310h47m12.585089842s
I191121 06:47:12.841582 131094 storage/replica_consistency.go:256  [n1,consistencyChecker,s1,r4/1:/System{/tsd-tse}] triggering stats recomputation to resolve delta of {ContainsEstimates:1394 LastUpdateNanos:1574318830864736188 IntentAge:0 GCBytesAge:0 LiveBytes:-33887 LiveCount:-670 KeyBytes:-32476 KeyCount:-670 ValBytes:-1411 ValCount:-670 IntentBytes:0 IntentCount:0 SysBytes:0 SysCount:0}
I191121 06:47:13.054286 128960 gossip/gossip.go:1531  [n1] node has connected to cluster via gossip
I191121 06:47:13.054823 128960 storage/stores.go:259  [n1] wrote 0 node addresses to persistent storage
I191121 06:47:13.270731 128922 util/stop/stopper.go:539  quiescing
W191121 06:47:13.271888 131104 storage/intentresolver/intent_resolver.go:818  failed to gc transaction record: could not GC completed transaction anchored at /Table/53/1/1/0: node unavailable; try another peer
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x3ee297a]

goroutine 129008 [running]:
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).Recover(0xc001a40140, 0x635ada0, 0xc0032818f0)
	/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:182 +0x155
panic(0x4f73580, 0x8e047d0)
	/usr/local/go/src/runtime/panic.go:522 +0x1b5
github.com/cockroachdb/cockroach/pkg/storage/rangefeed.(*Filter).NeedVal(0x0, 0xc000dce358, 0x4, 0x8, 0x0, 0x0, 0x0, 0xc000d86510)
	/go/src/github.com/cockroachdb/cockroach/pkg/storage/rangefeed/filter.go:45 +0x9a
```

SHA: https://github.com/cockroachdb/cockroach/commits/85b76090c30a58e43f778818795783eaca3758ff

Parameters:
```
TAGS=
GOFLAGS=-race -parallel=2
```

To repro, try:

```
# Don't forget to check out a clean suitable branch and experiment with the
# stress invocation until the desired results present themselves. For example,
# using stress instead of stressrace and passing the '-p' stressflag which
# controls concurrency.
./scripts/gceworker.sh start && ./scripts/gceworker.sh mosh
cd ~/go/src/github.com/cockroachdb/cockroach && \
stdbuf -oL -eL \
make stressrace TESTS=TestCatchupScanOrdering PKG=github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl TESTTIMEOUT=5m STRESSFLAGS='-maxtime 20m -timeout 10m' 2>&1 | tee /tmp/stress.log
```

Failed test: https://teamcity.cockroachdb.com/viewLog.html?buildId=1605078&tab=buildLog

```

I191122 07:26:41.396128 127895 sql/event_log.go:132  [n1,client=127.0.0.1:43000,user=root] Event: "create_table", target: 53, info: {TableName:d.public.bank Statement:CREATE TABLE IF NOT EXISTS bank (id INT8 PRIMARY KEY, balance INT8, payload STRING, FAMILY (id, balance, payload)) User:root}
I191122 07:26:41.407097 128145 storage/replica_command.go:413  [n1,split,s1,r24/1:/{Table/28-Max}] initiating a split of this range at key /Table/53 [r25] (zone config)
I191122 07:26:41.446002 128254 workload/workloadsql/dataload.go:135  imported bank (0s, 10 rows)
I191122 07:26:41.471688 127895 sql/event_log.go:132  [n1,client=127.0.0.1:43000,user=root] Event: "set_cluster_setting", target: 0, info: {SettingName:kv.range_merge.queue_enabled Value:false User:root}
I191122 07:26:41.473119 128254 workload/workloadsql/workloadsql.go:97  starting 9 splits
I191122 07:26:41.479081 127895 storage/replica_command.go:413  [n1,s1,r25/1:/{Table/53-Max}] initiating a split of this range at key /Table/53/1/5 [r26] (manual)
I191122 07:26:41.555801 127895 storage/replica_command.go:413  [n1,s1,r26/1:/{Table/53/1/5-Max}] initiating a split of this range at key /Table/53/1/8 [r27] (manual)
I191122 07:26:41.590299 127895 storage/replica_command.go:413  [n1,s1,r27/1:/{Table/53/1/8-Max}] initiating a split of this range at key /Table/53/1/9 [r28] (manual)
I191122 07:26:41.592016 128866 storage/replica_command.go:413  [n1,s1,r25/1:/Table/53{-/1/5}] initiating a split of this range at key /Table/53/1/3 [r29] (manual)
I191122 07:26:41.633768 128406 storage/replica_command.go:413  [n1,s1,r26/1:/Table/53/1/{5-8}] initiating a split of this range at key /Table/53/1/7 [r30] (manual)
I191122 07:26:41.639458 127895 storage/replica_command.go:413  [n1,s1,r29/1:/Table/53/1/{3-5}] initiating a split of this range at key /Table/53/1/4 [r31] (manual)
I191122 07:26:41.649449 128866 storage/replica_command.go:413  [n1,s1,r25/1:/Table/53{-/1/3}] initiating a split of this range at key /Table/53/1/2 [r32] (manual)
I191122 07:26:41.695579 127895 storage/replica_command.go:413  [n1,s1,r26/1:/Table/53/1/{5-7}] initiating a split of this range at key /Table/53/1/6 [r33] (manual)
I191122 07:26:41.711545 128866 storage/replica_command.go:413  [n1,s1,r25/1:/Table/53{-/1/2}] initiating a split of this range at key /Table/53/1/1 [r34] (manual)
I191122 07:26:42.617836 129264 ccl/changefeedccl/changefeed_processors.go:659  [n1,client=127.0.0.1:43012,user=root] sinkless feed span /Table/53/1{-/1} is behind by 437335h26m42.617691101s
I191122 07:26:42.781363 127325 gossip/gossip.go:1531  [n1] node has connected to cluster via gossip
I191122 07:26:42.781984 127325 storage/stores.go:259  [n1] wrote 0 node addresses to persistent storage
I191122 07:26:43.314509 127273 util/stop/stopper.go:539  quiescing
W191122 07:26:43.315342 129541 storage/intentresolver/intent_resolver.go:818  failed to gc transaction record: could not GC completed transaction anchored at /Table/53/1/1/0: node unavailable; try another peer
E191122 07:26:43.315755 129678 kv/txn_interceptor_committer.go:375  making txn commit explicit failed for "sql txn" meta={id=9810bbe7 key=/Table/53/1/0/0 pri=0.00603870 epo=0 ts=1574407603.305217286,0 min=1574407603.305217286,0 seq=3} rw=true stat=STAGING rts=1574407603.305217286,0 wto=false max=1574407603.305217286,0 int=1 ifw=1: result is ambiguous (server shutdown)
I191122 07:26:43.315859 127462 kv/transport_race.go:108  transport race promotion: ran 89 iterations on up to 1259 requests
W191122 07:26:43.318727 129282 internal/client/txn.go:518  [n1,client=127.0.0.1:43012,user=root] failure aborting transaction: node unavailable; try another peer; abort caused by: context canceled
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x3efb9ba]

goroutine 127419 [running]:
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).Recover(0xc0003d6500, 0x638d100, 0xc0030b7020)
	/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:182 +0x155
panic(0x4f98ca0, 0x8e4b7d0)
```

SHA: https://github.com/cockroachdb/cockroach/commits/1179c76e34d057259854d88c7948830d2b4b05cd

Parameters:
```
TAGS=
GOFLAGS=-race -parallel=2
```

To repro, try:

```
# Don't forget to check out a clean suitable branch and experiment with the
# stress invocation until the desired results present themselves. For example,
# using stress instead of stressrace and passing the '-p' stressflag which
# controls concurrency.
./scripts/gceworker.sh start && ./scripts/gceworker.sh mosh
cd ~/go/src/github.com/cockroachdb/cockroach && \
stdbuf -oL -eL \
make stressrace TESTS=TestCatchupScanOrdering PKG=github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl TESTTIMEOUT=5m STRESSFLAGS='-maxtime 20m -timeout 10m' 2>&1 | tee /tmp/stress.log
```

Failed test: https://teamcity.cockroachdb.com/viewLog.html?buildId=1607007&tab=buildLog

```

I191123 06:21:12.838689 138233 sql/event_log.go:132  [n1,client=127.0.0.1:36616,user=root] Event: "create_table", target: 53, info: {TableName:d.public.bank Statement:CREATE TABLE IF NOT EXISTS bank (id INT8 PRIMARY KEY, balance INT8, payload STRING, FAMILY (id, balance, payload)) User:root}
I191123 06:21:12.856248 138456 storage/replica_command.go:413  [n1,split,s1,r24/1:/{Table/28-Max}] initiating a split of this range at key /Table/53 [r25] (zone config)
I191123 06:21:12.872689 138327 workload/workloadsql/dataload.go:135  imported bank (0s, 10 rows)
I191123 06:21:12.899005 138233 sql/event_log.go:132  [n1,client=127.0.0.1:36616,user=root] Event: "set_cluster_setting", target: 0, info: {SettingName:kv.range_merge.queue_enabled Value:false User:root}
I191123 06:21:12.899955 138327 workload/workloadsql/workloadsql.go:97  starting 9 splits
I191123 06:21:12.904579 138233 storage/replica_command.go:413  [n1,s1,r25/1:/{Table/53-Max}] initiating a split of this range at key /Table/53/1/5 [r26] (manual)
I191123 06:21:12.965097 138233 storage/replica_command.go:413  [n1,s1,r25/1:/Table/53{-/1/5}] initiating a split of this range at key /Table/53/1/3 [r27] (manual)
I191123 06:21:12.991379 138233 storage/replica_command.go:413  [n1,s1,r25/1:/Table/53{-/1/3}] initiating a split of this range at key /Table/53/1/2 [r28] (manual)
I191123 06:21:12.996450 139042 storage/replica_command.go:413  [n1,s1,r26/1:/{Table/53/1/5-Max}] initiating a split of this range at key /Table/53/1/8 [r29] (manual)
I191123 06:21:13.027911 139031 storage/replica_command.go:413  [n1,s1,r27/1:/Table/53/1/{3-5}] initiating a split of this range at key /Table/53/1/4 [r30] (manual)
I191123 06:21:13.029597 138233 storage/replica_command.go:413  [n1,s1,r25/1:/Table/53{-/1/2}] initiating a split of this range at key /Table/53/1/1 [r31] (manual)
I191123 06:21:13.037716 139042 storage/replica_command.go:413  [n1,s1,r29/1:/{Table/53/1/8-Max}] initiating a split of this range at key /Table/53/1/9 [r32] (manual)
I191123 06:21:13.075391 139067 storage/replica_command.go:413  [n1,s1,r26/1:/Table/53/1/{5-8}] initiating a split of this range at key /Table/53/1/7 [r33] (manual)
I191123 06:21:13.107810 139031 storage/replica_command.go:413  [n1,s1,r26/1:/Table/53/1/{5-7}] initiating a split of this range at key /Table/53/1/6 [r34] (manual)
I191123 06:21:13.784519 139460 ccl/changefeedccl/changefeed_processors.go:659  [n1,client=127.0.0.1:36628,user=root] sinkless feed span /Table/53/1{-/3} is behind by 437358h21m13.784385841s
I191123 06:21:14.375283 137449 util/stop/stopper.go:539  quiescing
W191123 06:21:14.376740 139792 storage/intentresolver/intent_resolver.go:818  failed to gc transaction record: could not GC completed transaction anchored at /Table/53/1/8/0: node unavailable; try another peer
E191123 06:21:14.376861 139874 kv/txn_interceptor_committer.go:375  making txn commit explicit failed for "sql txn" meta={id=94f95507 key=/Table/53/1/4/0 pri=0.00400734 epo=0 ts=1574490074.366027569,0 min=1574490074.366027569,0 seq=3} rw=true stat=STAGING rts=1574490074.366027569,0 wto=false max=1574490074.366027569,0 int=1 ifw=1: result is ambiguous (server shutdown)
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x3eda29a]

goroutine 137551 [running]:
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).Recover(0xc0003e0460, 0x6343500, 0xc0022446c0)
	/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:182 +0x155
panic(0x4f77fe0, 0x8daf7d0)
	/usr/local/go/src/runtime/panic.go:522 +0x1b5
github.com/cockroachdb/cockroach/pkg/storage/rangefeed.(*Filter).NeedPrevVal(0x0, 0xc002f22cf8, 0x4, 0x8, 0x0, 0x0, 0x0, 0x52da5c0)
	/go/src/github.com/cockroachdb/cockroach/pkg/storage/rangefeed/filter.go:47 +0x9a
github.com/cockroachdb/cockroach/pkg/storage.(*Replica).populatePrevValsInLogicalOpLogRaftMuLocked(0xc001f42900, 0x6343500, 0xc002b38300, 0xc0003ddd80, 0x638b340, 0xc001124300)
	/go/src/github.com/cockroachdb/cockroach/pkg/storage/replica_rangefeed.go:466 +0x31d
```

SHA: https://github.com/cockroachdb/cockroach/commits/4de1b957fbda6eaa42c8fd7b0fb3278365f7a615

Parameters:
```
TAGS=
GOFLAGS=-race -parallel=2
```

To repro, try:

```
# Don't forget to check out a clean suitable branch and experiment with the
# stress invocation until the desired results present themselves. For example,
# using stress instead of stressrace and passing the '-p' stressflag which
# controls concurrency.
./scripts/gceworker.sh start && ./scripts/gceworker.sh mosh
cd ~/go/src/github.com/cockroachdb/cockroach && \
stdbuf -oL -eL \
make stressrace TESTS=TestCatchupScanOrdering PKG=github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl TESTTIMEOUT=5m STRESSFLAGS='-maxtime 20m -timeout 10m' 2>&1 | tee /tmp/stress.log
```

Failed test: https://teamcity.cockroachdb.com/viewLog.html?buildId=1607792&tab=buildLog

```

I191124 06:32:06.588236 125714 sql/event_log.go:132  [n1,client=127.0.0.1:44886,user=root] Event: "create_table", target: 53, info: {TableName:d.public.bank Statement:CREATE TABLE IF NOT EXISTS bank (id INT8 PRIMARY KEY, balance INT8, payload STRING, FAMILY (id, balance, payload)) User:root}
I191124 06:32:06.596436 125927 storage/replica_command.go:413  [n1,split,s1,r24/1:/{Table/28-Max}] initiating a split of this range at key /Table/53 [r25] (zone config)
I191124 06:32:06.622081 125736 workload/workloadsql/dataload.go:135  imported bank (0s, 10 rows)
I191124 06:32:06.642720 125714 sql/event_log.go:132  [n1,client=127.0.0.1:44886,user=root] Event: "set_cluster_setting", target: 0, info: {SettingName:kv.range_merge.queue_enabled Value:false User:root}
I191124 06:32:06.643272 125736 workload/workloadsql/workloadsql.go:97  starting 9 splits
I191124 06:32:06.646949 125714 storage/replica_command.go:413  [n1,s1,r25/1:/{Table/53-Max}] initiating a split of this range at key /Table/53/1/5 [r26] (manual)
I191124 06:32:06.707018 125714 storage/replica_command.go:413  [n1,s1,r26/1:/{Table/53/1/5-Max}] initiating a split of this range at key /Table/53/1/8 [r27] (manual)
I191124 06:32:06.736711 125714 storage/replica_command.go:413  [n1,s1,r26/1:/Table/53/1/{5-8}] initiating a split of this range at key /Table/53/1/7 [r28] (manual)
I191124 06:32:06.743170 126500 storage/replica_command.go:413  [n1,s1,r25/1:/Table/53{-/1/5}] initiating a split of this range at key /Table/53/1/3 [r29] (manual)
I191124 06:32:06.778273 126125 storage/replica_command.go:413  [n1,s1,r27/1:/{Table/53/1/8-Max}] initiating a split of this range at key /Table/53/1/9 [r30] (manual)
I191124 06:32:06.786999 125714 storage/replica_command.go:413  [n1,s1,r26/1:/Table/53/1/{5-7}] initiating a split of this range at key /Table/53/1/6 [r31] (manual)
I191124 06:32:06.792312 126500 storage/replica_command.go:413  [n1,s1,r29/1:/Table/53/1/{3-5}] initiating a split of this range at key /Table/53/1/4 [r32] (manual)
I191124 06:32:06.836188 126536 storage/replica_command.go:413  [n1,s1,r25/1:/Table/53{-/1/3}] initiating a split of this range at key /Table/53/1/2 [r33] (manual)
I191124 06:32:06.863694 126125 storage/replica_command.go:413  [n1,s1,r25/1:/Table/53{-/1/2}] initiating a split of this range at key /Table/53/1/1 [r34] (manual)
I191124 06:32:07.508465 126947 ccl/changefeedccl/changefeed_processors.go:659  [n1,client=127.0.0.1:44898,user=root] sinkless feed span /Table/53/1{-/2} is behind by 437382h32m7.508352859s
I191124 06:32:08.062827 124897 util/stop/stopper.go:539  quiescing
W191124 06:32:08.063769 126921 internal/client/txn.go:518  [n1,client=127.0.0.1:44898,user=root] failure aborting transaction: node unavailable; try another peer; abort caused by: context canceled
W191124 06:32:08.064599 127248 storage/intentresolver/intent_resolver.go:610  failed to cleanup transaction intents: failed to resolve intents: result is ambiguous (server shutdown)
I191124 06:32:08.065411 124800 kv/transport_race.go:108  transport race promotion: ran 21 iterations on up to 1274 requests
E191124 06:32:08.066166 127319 kv/txn_interceptor_committer.go:375  making txn commit explicit failed for "sql txn" meta={id=6ae17494 key=/Table/53/1/2/0 pri=0.02761143 epo=0 ts=1574577128.054894850,0 min=1574577128.054894850,0 seq=3} rw=true stat=STAGING rts=1574577128.054894850,0 wto=false max=1574577128.054894850,0 int=1 ifw=1: result is ambiguous (server shutdown)
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x3eda29a]

goroutine 125010 [running]:
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).Recover(0xc0017e0140, 0x6343500, 0xc00185ddd0)
	/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:182 +0x155
panic(0x4f77fe0, 0x8daf7d0)
	/usr/local/go/src/runtime/panic.go:522 +0x1b5
github.com/cockroachdb/cockroach/pkg/storage/rangefeed.(*Filter).NeedPrevVal(0x0, 0xc000dcf198, 0x4, 0x8, 0x0, 0x0, 0x0, 0x52da5c0)
```

SHA: https://github.com/cockroachdb/cockroach/commits/60b7b3b2ca511cf9adb30799fcf2b8cffd6223d9

Parameters:
```
TAGS=
GOFLAGS=-race -parallel=2
```

To repro, try:

```
# Don't forget to check out a clean suitable branch and experiment with the
# stress invocation until the desired results present themselves. For example,
# using stress instead of stressrace and passing the '-p' stressflag which
# controls concurrency.
./scripts/gceworker.sh start && ./scripts/gceworker.sh mosh
cd ~/go/src/github.com/cockroachdb/cockroach && \
stdbuf -oL -eL \
make stressrace TESTS=TestCatchupScanOrdering PKG=github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl TESTTIMEOUT=5m STRESSFLAGS='-maxtime 20m -timeout 10m' 2>&1 | tee /tmp/stress.log
```

Failed test: https://teamcity.cockroachdb.com/viewLog.html?buildId=1608663&tab=buildLog

```

I191125 07:10:50.428171 160759 sql/event_log.go:132  [n1,client=127.0.0.1:36744,user=root] Event: "create_table", target: 53, info: {TableName:d.public.bank Statement:CREATE TABLE IF NOT EXISTS bank (id INT8 PRIMARY KEY, balance INT8, payload STRING, FAMILY (id, balance, payload)) User:root}
I191125 07:10:50.442022 161023 storage/replica_command.go:413  [n1,split,s1,r24/1:/{Table/28-Max}] initiating a split of this range at key /Table/53 [r25] (zone config)
I191125 07:10:50.479550 160817 workload/workloadsql/dataload.go:135  imported bank (0s, 10 rows)
I191125 07:10:50.508383 160759 sql/event_log.go:132  [n1,client=127.0.0.1:36744,user=root] Event: "set_cluster_setting", target: 0, info: {SettingName:kv.range_merge.queue_enabled Value:false User:root}
I191125 07:10:50.509326 160817 workload/workloadsql/workloadsql.go:97  starting 9 splits
I191125 07:10:50.514481 160759 storage/replica_command.go:413  [n1,s1,r25/1:/{Table/53-Max}] initiating a split of this range at key /Table/53/1/5 [r26] (manual)
I191125 07:10:50.572433 160759 storage/replica_command.go:413  [n1,s1,r25/1:/Table/53{-/1/5}] initiating a split of this range at key /Table/53/1/3 [r27] (manual)
I191125 07:10:50.606072 161355 storage/replica_command.go:413  [n1,s1,r26/1:/{Table/53/1/5-Max}] initiating a split of this range at key /Table/53/1/8 [r28] (manual)
I191125 07:10:50.606889 160759 storage/replica_command.go:413  [n1,s1,r27/1:/Table/53/1/{3-5}] initiating a split of this range at key /Table/53/1/4 [r29] (manual)
I191125 07:10:50.645004 161360 storage/replica_command.go:413  [n1,s1,r25/1:/Table/53{-/1/3}] initiating a split of this range at key /Table/53/1/2 [r30] (manual)
I191125 07:10:50.650099 161355 storage/replica_command.go:413  [n1,s1,r28/1:/{Table/53/1/8-Max}] initiating a split of this range at key /Table/53/1/9 [r31] (manual)
I191125 07:10:50.656992 160759 storage/replica_command.go:413  [n1,s1,r26/1:/Table/53/1/{5-8}] initiating a split of this range at key /Table/53/1/7 [r32] (manual)
I191125 07:10:50.686243 161360 storage/replica_command.go:413  [n1,s1,r25/1:/Table/53{-/1/2}] initiating a split of this range at key /Table/53/1/1 [r33] (manual)
I191125 07:10:50.694948 160759 storage/replica_command.go:413  [n1,s1,r26/1:/Table/53/1/{5-7}] initiating a split of this range at key /Table/53/1/6 [r34] (manual)
I191125 07:10:51.475865 161956 ccl/changefeedccl/changefeed_processors.go:659  [n1,client=127.0.0.1:36756,user=root] sinkless feed span /Table/53/1{-/6} is behind by 437407h10m51.47575313s
I191125 07:10:51.630551 160057 gossip/gossip.go:1531  [n1] node has connected to cluster via gossip
I191125 07:10:51.630977 160057 storage/stores.go:259  [n1] wrote 0 node addresses to persistent storage
I191125 07:10:52.051518 160018 util/stop/stopper.go:539  quiescing
W191125 07:10:52.053029 162337 storage/intentresolver/intent_resolver.go:610  failed to cleanup transaction intents: failed to resolve intents: result is ambiguous (server shutdown)
E191125 07:10:52.053210 162254 kv/txn_interceptor_committer.go:375  making txn commit explicit failed for "sql txn" meta={id=1d689bcb key=/Table/53/1/0/0 pri=0.01012636 epo=0 ts=1574665852.043800131,0 min=1574665852.043800131,0 seq=3} rw=true stat=STAGING rts=1574665852.043800131,0 wto=false max=1574665852.043800131,0 int=1 ifw=1: result is ambiguous (server shutdown)
W191125 07:10:52.054797 161841 internal/client/txn.go:518  [n1,client=127.0.0.1:36756,user=root] failure aborting transaction: node unavailable; try another peer; abort caused by: context canceled
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x3ed929a]

goroutine 160085 [running]:
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).Recover(0xc000ae8140, 0x63416c0, 0xc004acc690)
	/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:182 +0x155
panic(0x4f76b40, 0x8dac7d0)
	/usr/local/go/src/runtime/panic.go:522 +0x1b5
```

SHA: https://github.com/cockroachdb/cockroach/commits/78f8482fd864376a26f32dd983afeb679d1c5977

Parameters:
```
TAGS=
GOFLAGS=-race -parallel=2
```

To repro, try:

```
# Don't forget to check out a clean suitable branch and experiment with the
# stress invocation until the desired results present themselves. For example,
# using stress instead of stressrace and passing the '-p' stressflag which
# controls concurrency.
./scripts/gceworker.sh start && ./scripts/gceworker.sh mosh
cd ~/go/src/github.com/cockroachdb/cockroach && \
stdbuf -oL -eL \
make stressrace TESTS=TestCatchupScanOrdering PKG=github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl TESTTIMEOUT=5m STRESSFLAGS='-maxtime 20m -timeout 10m' 2>&1 | tee /tmp/stress.log
```

Failed test: https://teamcity.cockroachdb.com/viewLog.html?buildId=1611226&tab=buildLog

```

I191126 06:44:33.203498 136990 sql/event_log.go:132  [n1,client=127.0.0.1:49450,user=root] Event: "create_table", target: 53, info: {TableName:d.public.bank Statement:CREATE TABLE IF NOT EXISTS bank (id INT8 PRIMARY KEY, balance INT8, payload STRING, FAMILY (id, balance, payload)) User:root}
I191126 06:44:33.213074 137318 storage/replica_command.go:413  [n1,split,s1,r24/1:/{Table/28-Max}] initiating a split of this range at key /Table/53 [r25] (zone config)
I191126 06:44:33.239928 137205 workload/workloadsql/dataload.go:135  imported bank (0s, 10 rows)
I191126 06:44:33.258665 136990 sql/event_log.go:132  [n1,client=127.0.0.1:49450,user=root] Event: "set_cluster_setting", target: 0, info: {SettingName:kv.range_merge.queue_enabled Value:false User:root}
I191126 06:44:33.259290 137205 workload/workloadsql/workloadsql.go:97  starting 9 splits
I191126 06:44:33.262992 136990 storage/replica_command.go:413  [n1,s1,r25/1:/{Table/53-Max}] initiating a split of this range at key /Table/53/1/5 [r26] (manual)
I191126 06:44:33.312948 136990 storage/replica_command.go:413  [n1,s1,r25/1:/Table/53{-/1/5}] initiating a split of this range at key /Table/53/1/3 [r27] (manual)
I191126 06:44:33.341719 136990 storage/replica_command.go:413  [n1,s1,r27/1:/Table/53/1/{3-5}] initiating a split of this range at key /Table/53/1/4 [r28] (manual)
I191126 06:44:33.349889 137855 storage/replica_command.go:413  [n1,s1,r26/1:/{Table/53/1/5-Max}] initiating a split of this range at key /Table/53/1/8 [r29] (manual)
I191126 06:44:33.384986 137855 storage/replica_command.go:413  [n1,s1,r26/1:/Table/53/1/{5-8}] initiating a split of this range at key /Table/53/1/7 [r30] (manual)
I191126 06:44:33.387956 137681 storage/replica_command.go:413  [n1,s1,r25/1:/Table/53{-/1/3}] initiating a split of this range at key /Table/53/1/2 [r31] (manual)
I191126 06:44:33.393713 136990 storage/replica_command.go:413  [n1,s1,r29/1:/{Table/53/1/8-Max}] initiating a split of this range at key /Table/53/1/9 [r32] (manual)
I191126 06:44:33.415910 137855 storage/replica_command.go:413  [n1,s1,r26/1:/Table/53/1/{5-7}] initiating a split of this range at key /Table/53/1/6 [r33] (manual)
I191126 06:44:33.424203 137681 storage/replica_command.go:413  [n1,s1,r25/1:/Table/53{-/1/2}] initiating a split of this range at key /Table/53/1/1 [r34] (manual)
I191126 06:44:34.058080 138113 ccl/changefeedccl/changefeed_processors.go:659  [n1,client=127.0.0.1:49458,user=root] sinkless feed span /Table/53/1{-/1} is behind by 437430h44m34.057952919s
I191126 06:44:34.543384 136337 gossip/gossip.go:1531  [n1] node has connected to cluster via gossip
I191126 06:44:34.543743 136337 storage/stores.go:259  [n1] wrote 0 node addresses to persistent storage
I191126 06:44:34.657613 136295 util/stop/stopper.go:539  quiescing
W191126 06:44:34.658762 138458 storage/intentresolver/intent_resolver.go:818  failed to gc transaction record: could not GC completed transaction anchored at /Table/53/1/5/0: node unavailable; try another peer
E191126 06:44:34.661646 138695 kv/txn_interceptor_committer.go:375  making txn commit explicit failed for "sql txn" meta={id=8d66583c key=/Table/53/1/7/0 pri=0.02648220 epo=0 ts=1574750674.638832720,0 min=1574750674.638832720,0 seq=3} rw=true stat=STAGING rts=1574750674.638832720,0 wto=false max=1574750674.638832720,0 int=1 ifw=1: result is ambiguous (server shutdown)
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x3edab4a]

goroutine 136421 [running]:
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).Recover(0xc0003e6320, 0x634cb00, 0xc002c25620)
	/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:182 +0x155
panic(0x4f7ac80, 0x8dba7c0)
	/usr/local/go/src/runtime/panic.go:522 +0x1b5
github.com/cockroachdb/cockroach/pkg/storage/rangefeed.(*Filter).NeedPrevVal(0x0, 0xc001e57678, 0x4, 0x8, 0x0, 0x0, 0x0, 0x52dd460)
	/go/src/github.com/cockroachdb/cockroach/pkg/storage/rangefeed/filter.go:47 +0x9a
```

--- a/pkg/storage/rangefeed/processor.go
+++ b/pkg/storage/rangefeed/processor.go
@@ -391,7 +391,7 @@ func (p *Processor) Len() int {
 }
 
 // Filter returns a new operation filter based on the registrations attached to
-// the processor.
+// the processor. Returns nil if the processor has been stopped already.
 func (p *Processor) Filter() *Filter {
 	if p == nil {
 		return nil

--- a/pkg/storage/replica_rangefeed.go
+++ b/pkg/storage/replica_rangefeed.go
@@ -254,6 +254,7 @@ func (r *Replica) setRangefeedProcessor(p *rangefeed.Processor) {
 	r.rangefeedMu.Lock()
 	defer r.rangefeedMu.Unlock()
 	r.rangefeedMu.proc = p
+	r.updateRangefeedFilterLocked()
 	r.store.addReplicaWithRangefeed(r.RangeID)
 }
 
@@ -370,9 +371,6 @@ func (r *Replica) registerWithRangefeedRaftMuLocked(
 	// process could have raced with ours because calling this method
 	// requires raftMu to be exclusively locked.
 	r.setRangefeedProcessor(p)
-
-	// Set the rangefeed filter now that the processor is set.
-	r.updateRangefeedFilterLocked()
 
 	// Check for an initial closed timestamp update immediately to help
 	// initialize the rangefeed's resolved timestamp as soon as possible.

--- a/pkg/storage/replica_rangefeed.go
+++ b/pkg/storage/replica_rangefeed.go
@@ -254,7 +254,11 @@ func (r *Replica) setRangefeedProcessor(p *rangefeed.Processor) {
 	r.rangefeedMu.Lock()
 	defer r.rangefeedMu.Unlock()
 	r.rangefeedMu.proc = p
-	r.updateRangefeedFilterLocked()
+	if !r.updateRangefeedFilterLocked() {
+		// This can't happen. We just set the processor and haven't released the
+		// exclusive lock, so no other goroutine could have stopped it.
+		panic("rangefeed processor unexpectedly stopped")
+	}
 	r.store.addReplicaWithRangefeed(r.RangeID)
 }
 
@@ -274,8 +278,12 @@ func (r *Replica) unsetRangefeedProcessor(p *rangefeed.Processor) {
 	r.unsetRangefeedProcessorLocked(p)
 }
 
-func (r *Replica) updateRangefeedFilterLocked() {
+func (r *Replica) updateRangefeedFilterLocked() bool {
 	r.rangefeedMu.opFilter = r.rangefeedMu.proc.Filter()
+	// Return whether the update to the filter was successful or not. If
+	// the processor was already stopped then we can't update the filter.
+	stopped := r.rangefeedMu.opFilter == nil
+	return !stopped
 }
 
 // The size of an event is 112 bytes, so this will result in an allocation on
@@ -307,11 +315,10 @@ func (r *Replica) registerWithRangefeedRaftMuLocked(
 	p := r.rangefeedMu.proc
 	if p != nil {
 		reg := p.Register(span, startTS, catchupIter, withDiff, stream, errC)
-		if reg {
+		if reg && r.updateRangefeedFilterLocked() {
 			// Registered successfully with an existing processor.
 			// Update the rangefeed filter to avoid filtering ops
 			// that this new registration might be interested in.
-			r.updateRangefeedFilterLocked()
 			r.rangefeedMu.Unlock()
 			return p
 		}
@@ -388,13 +395,11 @@ func (r *Replica) maybeDisconnectEmptyRangefeed(p *rangefeed.Processor) {
 		// The processor has already been removed or replaced.
 		return
 	}
-	if p.Len() == 0 {
-		// Stop the rangefeed processor.
+	if p.Len() == 0 || !r.updateRangefeedFilterLocked() {
+		// Stop the rangefeed processor if it has no registrations or if we are
+		// unable to update the operation filter.
 		p.Stop()
 		r.unsetRangefeedProcessorLocked(p)
-	} else {
-		// Update the rangefeed filter to filter more aggressively.
-		r.updateRangefeedFilterLocked()
 	}
 }
 

